src/DrDotnet.Profilers/profilers/merged_call_stacks_profiler.rs

```diff
@@ -0,0 +1,258 @@
+
+use crate::api::*;
+use crate::api::ffi::{FunctionID, HRESULT, ThreadID};
+use crate::macros::*;
+use crate::profilers::*;
+use crate::session::Report;
+
+const PADDING: usize = 5;
+const NB_THREAD_IDS_TO_PRINT: usize = 4;
+
+
+#[derive(Default)]
+pub struct MergedCallStacksProfiler {
+    clr_profiler_info: ClrProfilerInfo,
+    session_info: SessionInfo,
+    merged_stack: Arc<Mutex<MergedStack>>
+}
+
+#[derive(Default, Debug, Eq, PartialEq, Hash, Clone)]
+enum StackFrameType {
+    #[default]
+    Managed,
+    Native
+}
+
+#[derive(Default, Clone, Debug)]
+struct StackFrame {
+    display: Option<String>,
+    kind: StackFrameType,
+    fct_id: FunctionID,
+}
+
+#[derive(Default, Clone, Debug)]
+struct MergedStack {
+    thread_ids: Vec<ThreadID>,
+    stacks: Vec<MergedStack>,
+    frame: StackFrame
+}
+
+impl StackFrame {
+    fn format(frame: &StackFrame, clr: &ClrProfilerInfo) -> String {
+        match frame.kind {
+            StackFrameType::Native => "unmanaged".to_string(),
+            StackFrameType::Managed =>  unsafe { clr.get_full_method_name(frame.fct_id) }
+        }
+    }
+}
+
+impl MergedStack {
+    
+    pub fn push_thread_id(&mut self, thread_id: ThreadID) {
+        self.thread_ids.push(thread_id);
+    }
+    
+    pub fn add_stack(&mut self, clr: &ClrProfilerInfo, thread_id: ThreadID, stack_trace: Vec<StackFrame>, index: Option<usize>) {
+        self.thread_ids.push(thread_id);
+        
+        let mut index = index.unwrap_or(0);
+
+        let mut first_frame: Option<&StackFrame> = None;
+        
+        for frame in &stack_trace[index..] {
+            if frame.kind == StackFrameType::Managed { 
+                first_frame = Some(frame);
+                break
+            }
+            index += 1;
+        }
+        
+        if first_frame.is_none() {
+            return; 
+        }
+        let first_frame = first_frame.unwrap();
+        
+        let mut merged_stack = self.stacks.iter_mut()
+            .find_or_first(|s| s.frame.fct_id == first_frame.fct_id);
+        
+        if merged_stack.is_none() {
+            self.stacks.push(MergedStack::new(&first_frame, clr));
+            merged_stack = self.stacks.last_mut();
+        }
+        
+        let merged_stack = merged_stack.unwrap();
+        if index == stack_trace.len() -1 {
+            merged_stack.push_thread_id(thread_id);
+        } else {
+            merged_stack.add_stack(clr, thread_id, stack_trace, Some(index + 1));
+        }
+    }
+    
+    pub fn new(frame: &StackFrame, clr: &ClrProfilerInfo) -> Self {
+        MergedStack {
+            thread_ids: Vec::new(),
+            stacks: Vec::new(),
+            frame: StackFrame {
+                kind: frame.kind.clone(),
+                fct_id: frame.fct_id,
+                display: Some(StackFrame::format(frame, clr))
+            }
+        }
+    }
+
+    pub fn write_to(&self, report: &mut Report) {
+        self.write_stack(report, 0);
```

Comment:
```
# Merged Callstacks

                                              ~~~~ 
                                                 1 System.Threading.ManualResetEventSlim.Wait
                                                 1 System.Threading.ManualResetEventSlim.Wait
                                                 1 System.Net.Sockets.SocketAsyncContext.PerformSyncOperation
                                                 1 System.Net.Sockets.SocketAsyncContext.ReceiveFrom
                                            2 System.Threading.ManualResetEventSlim.Wait
                                            2 System.Threading.ManualResetEventSlim.Wait
                                            2 NUnit.Framework.Api.NUnitTestAssemblyRunner.WaitForCompletion
                                            2 NUnit.Framework.Api.NUnitTestAssemblyRunner.Run
                                            2 NUnit.Framework.Api.FrameworkController.RunTests
                                       3 Sys.<ReceiveMessage>g____PInvoke|60_0
                                  4 System.Threading.ManualResetEventSlim.Wait
                             5 System.Threading.ManualResetEventSlim.Wait
                        6 System.Threading.Tasks.Task.SpinThenBlockingWait
                        6 System.Threading.Tasks.Task.InternalWaitCore
                        6 System.Threading.Tasks.Task.Wait
                        6 System.Threading.Tasks.Task.Wait
                        6 JetBrains.ReSharper.TestRunner.Implementation.UnitTestRemoteAgent.ConnectAndWaitForShutdown
                        6 JetBrains.ReSharper.TestRunner.Application.Run
                        6 System.Reflection.MethodInvoker.Invoke
                        6 System.Reflection.RuntimeMethodInfo.Invoke
                        6 JetBrains.ReSharper.TestRunner.Utils.MethodInvoker.InvokeOn
                        6 JetBrains.ReSharper.TestRunner.Extensions.LoadContextExtensions.RunInContext
                  10 JetBrains.ReSharper.TestRunner.Bootstrapper`1.RunAppInSeparateLoadContext
             13 JetBrains.ReSharper.TestRunner.Bootstrapper`1.Run
        14 JetBrains.ReSharper.TestRunner.Bootstrapper`1.Run
        14 CommandLine.ParserResultExtensions.MapResult
   19 JetBrains.ReSharper.TestRunner.EntryPoint.Run
   19 JetBrains.ReSharper.TestRunner.Program.Run
   19 JetBrains.ReSharper.TestRunner.Program.Main
==> 19 threads with 1 roots
```
Aren't we missing some branches? Because I don't see the 4 threads doing Fibonacci

Reply:
Right, good question. Will look into that